test/util.test.js

```diff
@@ -2,10 +2,345 @@ var assert = require('assert');
+      sub: {
+        enabled: false,
+        sub2: {
+          font: 'awesome'
+        }
+      }
+    };
+
+    assert(a.color !== undefined && a.color === result.color);
+    assert(a.notInSource === true);
+    if (checkCopyTarget) {
+      assert(a.notInTarget === true);
+    } else {
+      assert(a.notInTarget === undefined);
+    }
+
+    var sub = a.sub;
+    assert(sub !== undefined);
+    assert(sub.enabled !== undefined && sub.enabled === result.sub.enabled);
+    assert(sub.notInSource === true);
+    if (checkCopyTarget) {
+      assert(sub.notInTarget === true);
+    } else {
+      assert(sub.notInTarget === undefined);
+    }
+
+    sub = a.sub.sub2;
+    assert(sub !== undefined);
+    assert(sub !== undefined && sub.font !== undefined && sub.font === result.sub.sub2.font);
+    assert(sub.notInSource === true);
+    assert(a.subNotInSource !== undefined);
+    if (checkCopyTarget) {
+      assert(a.subNotInTarget.enabled === true);
+      assert(sub.notInTarget === true);
+    } else {
+      assert(a.subNotInTarget === undefined);
+      assert(sub.notInTarget === undefined);
+    }
+  }
+
+
+  /**
+   * Spot check on values of a unchanged as intended
+   */
+  function testAUnchanged(a) {
+    var sub = a.sub;
+    assert(sub !== undefined);
+    assert(sub.enabled !== undefined && sub.enabled === true);
+    assert(sub.notInSource === true);
+    assert(sub.notInTarget === undefined);
+    assert(sub.deleteThis === true);
+
+    sub = a.sub.sub2;
+    assert(sub !== undefined);
+    assert(sub !== undefined && sub.font !== undefined && sub.font === 'arial');
+    assert(sub.notInSource === true);
+    assert(sub.notInTarget === undefined);
+
+    assert(a.subNotInSource !== undefined);
+    assert(a.subNotInTarget === undefined);
+  }
+
+
+  function initA() {
+    return {
+      color: 'red',
+      notInSource: true,
+      sub: {
+        enabled: true,
+        notInSource: true,
+        sub2: {
+          font: 'arial',
+          notInSource: true,
+        },
+        deleteThis: true,
+      },
+      subNotInSource: {
+        enabled: true,
+      },
+      deleteThis: true,
+      subDeleteThis: {
+        enabled: true,
+      },
+    };
+  }
+
+
+  beforeEach(function() {
+    this.a = initA();
+
+    this.b = {
+      color: 'green',
+      notInTarget: true,
+      sub: {
+        enabled: false,
+        notInTarget: true,
+        sub2: {
+          font: 'awesome',
+          notInTarget: true,
+        },
+        deleteThis: null,
+      },
+      subNotInTarget: {
+        enabled: true,
+      },
+      deleteThis: null,
+      subDeleteThis: null
+    };
+  });
+
+
+  it('performs fillIfDefined() as advertized', function () {
+    var a = this.a;
+    var b = this.b;
+
+    util.fillIfDefined(a, b);
+    checkExtended(a, b);
+
+    // NOTE: if allowDeletion === false, null values are copied over!
+    //       This is due to existing logic; it might not be the intention and hence a bug
```

Comment:
Umm, I don't think it's a good behavior. It does seem like a bug that should be addressed. But on the other-hand, I usually take the approach that if a field is undefined, it should not be assigned a value of `undefined` whilst `null` is a legit value in a field and should be passed on in the extend. If that's the behavior of the function, I'm ok with it.

Reply:
I agree with you; for me the trade-off is between fixing it and keeping it like this so as not to break things elsewhere. I'll have a look if I can get rid of it, or confirm that the current working is good.